target/linux/bmips/files/drivers/net/ethernet/broadcom/bcm6368-enetsw.c

```diff
@@ -251,32 +254,35 @@ static inline void dmas_writel(struct bcm6368_enetsw *priv, u32 val,
 /*
  * refill rx queue
  */
-static int bcm6368_enetsw_refill_rx(struct net_device *dev)
+static int bcm6368_enetsw_refill_rx(struct net_device *dev, bool napi_mode)
 {
 	struct bcm6368_enetsw *priv = netdev_priv(dev);
 
 	while (priv->rx_desc_count < priv->rx_ring_size) {
 		struct bcm6368_enetsw_desc *desc;
-		struct sk_buff *skb;
-		dma_addr_t p;
 		int desc_idx;
 		u32 len_stat;
 
 		desc_idx = priv->rx_dirty_desc;
 		desc = &priv->rx_desc_cpu[desc_idx];
 
-		if (!priv->rx_skb[desc_idx]) {
-			skb = netdev_alloc_skb(dev, priv->rx_skb_size);
-			if (!skb)
+		if (!priv->rx_buf[desc_idx]) {
+			unsigned char *buf;
+
+			if(likely(napi_mode))
+				buf = napi_alloc_frag(priv->rx_frag_size);
+			else
+				buf = netdev_alloc_frag(priv->rx_frag_size);
+			if (unlikely(!buf))
 				break;
-			priv->rx_skb[desc_idx] = skb;
-			p = dma_map_single(&priv->pdev->dev, skb->data,
-					   priv->rx_skb_size,
-					   DMA_FROM_DEVICE);
-			desc->address = p;
+			priv->rx_buf[desc_idx] = buf;
+			desc->address = dma_map_single(&priv->pdev->dev,
+										   buf + NET_SKB_PAD,
```

Comment:
what happen here with the spacing?

Reply:
Sorry, vscode messes up the spacing sometimes.

The bcm63xx_net driver, which this driver is based on, doesn't check the dma addr.
I can add dma_mapping_error() checking if needed.

Reply:
yes pls add the dma check
Np about the space it was just to warn you on what to fix :D

Reply:
(if you want you can put the dma_mapping_error() in a separate commit)

Reply:
> (if you want you can put the dma_mapping_error() in a separate commit)

https://github.com/openwrt/openwrt/pull/10144/commits/5cf4fccdecc8f980e3bf3284781d2e013e4a0dba

Reply:
Done